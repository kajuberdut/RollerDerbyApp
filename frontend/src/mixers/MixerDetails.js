import React, { useState, useEffect } from "react";
import FastApi from "../Api";
import { useParams } from "react-router-dom";
import Loading from "../multiUse/loading/Loading";
import "./MixerDetails.css"
import { MDBCol, MDBContainer, MDBRow, MDBCard, MDBCardText, MDBCardBody } from 'mdb-react-ui-kit';

/**  
 * Display mixer detail page
 */

function MixerDetail({ getAllChats }) {

   /** Get eventId from url, set mixer, address and is loading in state*/

    const eventId = useParams(); 
    const [mixer, setMixer ] = useState([]);
    const [address, setAddress ] = useState([]);
    const [isLoading, setIsLoading] = useState(true);
    const [connected, setConnected] = useState(null)
    const [error, setError] = useState(null)

    /** Retrieve user from local storage*/

    const user = JSON.parse(localStorage.getItem('user'));

    /** API get request for a specific mixer */ 

    async function getMixer() {

      try {

        let mixer = await FastApi.getMixer(eventId.id);
        let address = await FastApi.getAddress(mixer.addressId)
        setMixer(mixer);
        setAddress(address);
        setIsLoading(false);
      } catch(errors) {

        return { success: false, errors };
      }
    }
   
    /** Reloading mixers when it changes request for mixers */
   
    useEffect(() => {
      getMixer();
    }, []);

    /** Display isLoading if API call is has not returned */

    if (isLoading) {
      return (
          <Loading />
      )
    }

    /** Handle click of button  */

    async function handleClick(e) {
      e.preventDefault(); 

      //   /** Retrieve user from local storage  */ 
      const user = JSON.parse(localStorage.getItem('user'));

      let data = {userId: `${user.userId}`, groupId: `${mixer.groupId}`}

      try {     
        let result = await FastApi.addUserToGroup(data); 

        if(result) {
          getAllChats();
        }

        setConnected("Connected. Check your chats.")
        return { success: true};

      } catch(errors) {
        setError("Unable to connect to chat.")
        return {success: false, description: "Unable to connect to chat."};
      }
    }

    return (

    <section>
        <div className="MixerDetails" style={{marginRight: '35%', marginTop: '150px', minWidth: '400px', marginLeft: '10%'}} >
          <MDBContainer>
            <MDBRow className="justify-content-center align-items-center h-100"> 
            <MDBCol lg="9" xl="10">
              <MDBCard style={{minWidth: '300px', marginTop: '50px', boxShadow: '0 2px 2px 0 rgba(0,0,0,0.14), 0 3px 1px -2px rgba(0,0,0,0.12), 0 1px 5px 0 rgba(0,0,0,0.2)'}}>
                <div className="rounded-top text-white d-flex flex-row" style={{ backgroundColor: '#000', height: '200px'}}>

                  <button type="button" className="btn btn-outline-dark" onClick={handleClick} data-mdb-ripple-color="dark"
                      style={{zIndex: 1, height: '40px', backgroundColor: '#d1d2d4', position: 'absolute', right: '20px', marginTop: '10px', fontSize: '15px'}}>
                      Join Chat 
                  </button>

                  {(error || connected) && (
                      <div style={{ color: error ? 'red' : 'green',fontSize: '14px', position: 'absolute',right: '10px', top: '40px', marginTop: '10px'}}
                      >
                        {error || connected}
                      </div>
                    )}

                  <div className="ms-3" style={{display: 'flex', overflow: 'hidden'}}>
                    <MDBCardText tag="h1" style={{ marginTop: '50px'}}>{mixer.theme}</MDBCardText>
                    <MDBCardText tag="h4" style={{position: 'absolute'}}>{mixer.date}</MDBCardText>
                  </div>
                </div>
                <div className="p-4 text-black" style={{ backgroundColor: '#f8f9fa' }}>
         
                  <div className="d-flex justify-content-beginning text-center py-1" style={{marginTop: '2px'}}>
             
                  <div className="d-flex flex-column text-center py-1">
                    <MDBCardText>{address.streetAddress}</MDBCardText>
                    <MDBCardText style={{ position: 'absolute', top: '16rem' }}>{address.city}, {address.state} {address.zipCode}</MDBCardText>
                  </div>
                  </div>
                    <div className="d-flex justify-content-end text-center py-1" style={{marginTop: '2px'}}>
                    <div>
                      <MDBCardText className="mb-1 h5" style={{marginRight: '30px'}}>{mixer.level}</MDBCardText>
                      <MDBCardText className="small text-muted mb-0"style={{marginRight: '30px'}}>level</MDBCardText>
                    </div>
  
                    <div className="px-3">
                      <MDBCardText className="mb-1 h6" style={{marginLeft: '30px'}}>{mixer.ruleset}</MDBCardText>
                      <MDBCardText className="small text-muted mb-0" style={{marginLeft: '30px', marginTop: '7px'}}>ruleset</MDBCardText>
                    </div>
                   { mixer.coEd && <div className="px-3">
                      <MDBCardText className="mb-1 h6" style={{marginLeft: '30px'}}>Yes</MDBCardText>
                      <MDBCardText className="small text-muted mb-0" style={{marginLeft: '30px', marginTop: '7px'}}>open</MDBCardText>
                    </div>
                    }       
                     { !mixer.coEd && <div className="px-3">
                      <MDBCardText className="mb-1 h6" style={{marginLeft: '30px'}}>No</MDBCardText>
                      <MDBCardText className="small text-muted mb-0" style={{marginLeft: '30px', marginTop: '7px'}}>open</MDBCardText>
                    </div>
                    }   
                  </div>
                </div>
                <MDBCardBody className="text-black p-4">
                <div className="mb-5">
                    <p className="lead fw-normal mb-1">Time</p>
                    <div className="p-4" style={{ backgroundColor: '#f8f9fa' }}>
                      <MDBCardText className="font-italic mb-1">{mixer.time}</MDBCardText>
                      <MDBCardText className="font-italic mb-1">{mixer.timeZone}</MDBCardText>
                    </div>
                  </div>
                  <div className="mb-5">
                    <p className="lead fw-normal mb-1">Signup Here</p>
                    <div className="p-4" style={{ backgroundColor: '#f8f9fa' }}>
                      <MDBCardText className="font-italic mb-1">{mixer.signupLink}</MDBCardText>
                    </div>
                  </div>
                  <div className="mb-5">
                    <p className="lead fw-normal mb-1">Floor Type</p>
                    <div className="p-4" style={{ backgroundColor: '#f8f9fa' }}>
                      <MDBCardText className="font-italic mb-1">{mixer.floorType}</MDBCardText>
                    </div>
                  </div>
                  <div className="mb-5">
                    <p className="lead fw-normal mb-1">Description</p>
                    <div className="p-4" style={{ backgroundColor: '#f8f9fa' }}>
                      <MDBCardText className="font-italic mb-1">{mixer.description}</MDBCardText>
                    </div>
                  </div>
                  <div className="mb-5">
                    <p className="lead fw-normal mb-1">Jersey Colors</p>
                    <div className="p-4" style={{ backgroundColor: '#f8f9fa' }}>
                      <MDBCardText className="font-italic mb-1">{mixer.jerseyColors}</MDBCardText>
                    </div>
                  </div>
                </MDBCardBody>
              </MDBCard>
            </MDBCol>
          </MDBRow>
        </MDBContainer>
        </div>
   </section>
  );
}

export default MixerDetail