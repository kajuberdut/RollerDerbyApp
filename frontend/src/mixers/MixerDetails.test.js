import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MixerDetails from './MixerDetails';
import FastApi from '../Api'; 
import { useParams } from 'react-router-dom';


//  Passing as of 2/13/24
// npm test MixerDetails.test.js
// must be in frontend directory 

// Mock FastApi methods
// This is where your error about mockResolvedValue was coming from
jest.mock('../Api', () => ({
  getMixer: jest.fn(),
  getAddress: jest.fn(),
  addUserToGroup: jest.fn(),
}));

// Mock useParams
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // import and spread the actual module
  useParams: jest.fn(),
}));

describe('MixerDetails Component', () => {
  const mockMixer = {
    eventId: 2,
		type: "mixer",
	  date: "2024-01-13",
		addressId: 2,
		time: "13:55",
		timeZone: " Mountain Time (MT): America/Denver (Denver, Phoenix, Salt Lake City)",
		theme: "HOLIDAY CHEER MIxer",
		description: "blah blah blah",
	  level: "AA",
		coEd: false,
		ruleset: "WFTDA",
		floorType: "polished concrete",
		jerseyColors: "pink and green",
		signupLink: "https://signup.com"
	}

  const mockAddress = {
    streetAddress: '123 Main St',
    city: 'Anytown',
    state: 'State',
    zipCode: '12345',
    // Add other necessary mock address details here
  };

  beforeEach(() => {
    FastApi.getMixer.mockResolvedValue(mockMixer);
    FastApi.getAddress.mockResolvedValue(mockAddress);
    FastApi.addUserToGroup.mockResolvedValue({ success: true });
    useParams.mockReturnValue({ id: '1' });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders mixer details page', () => {
    render(<MixerDetails />);
  });


  test('matches snapshot', function() {
    const { asFragment } = render(<MixerDetails />);
    expect(asFragment()).toMatchSnapshot(); 
  })

  test('renders loading when mixer details has not resolved', () => {
      
    render(<MixerDetails />);

    const loading = screen.getByTestId('loading-indicator');
    expect(loading).toBeInTheDocument();

});

it('fetches mixer and address details and displays theme and date', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText(mockMixer.theme)).toBeInTheDocument();
    expect(screen.getByText(mockMixer.date)).toBeInTheDocument();
  });
});

it('fetches mixer and address details and displays level', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText('level')).toBeInTheDocument();
    expect(screen.getByText(mockMixer.level)).toBeInTheDocument();
  });
});

it('fetches mixer and address details and displays ruleset', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText('ruleset')).toBeInTheDocument();
    expect(screen.getByText(mockMixer.ruleset)).toBeInTheDocument();
  });
});

it('fetches mixer and address details and displays time', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText('Time')).toBeInTheDocument();
    expect(screen.getByText(mockMixer.time)).toBeInTheDocument();
    });
});


it('fetches mixer and address details and displays signup', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText('Signup Here')).toBeInTheDocument();
    expect(screen.getByText(mockMixer.signupLink)).toBeInTheDocument();
  });
});

it('fetches mixer and address details and displays floor type', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText('Floor Type')).toBeInTheDocument();
    expect(screen.getByText(mockMixer.floorType)).toBeInTheDocument();
  });
});

it('fetches mixer and address details and displays floor type', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText('Floor Type')).toBeInTheDocument();
    expect(screen.getByText(mockMixer.floorType)).toBeInTheDocument();
  });

});

it('fetches mixer and address details and displays address details', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

  // Wait for the API call to resolve and check if the details are displayed
  await waitFor(() => {
    expect(screen.getByText(mockAddress.streetAddress)).toBeInTheDocument();
    expect(screen.getByText(`${mockAddress.city}, ${mockAddress.state} ${mockAddress.zipCode}`)).toBeInTheDocument();
  });

});

it('allows a user to join a chat and updates the UI on success', async () => {
  render(<MixerDetails getAllChats={() => {}} />);

  // Wait for the component to finish loading
  await waitFor(() => {
    expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
  });

  await waitFor(() => {
    // Simulate clicking the "Join Chat" button
      userEvent.click(screen.getByText(/join chat/i));
      // Check for the success message
      expect(screen.getByText(/connected. check your chats./i)).toBeInTheDocument();
  });
});


});