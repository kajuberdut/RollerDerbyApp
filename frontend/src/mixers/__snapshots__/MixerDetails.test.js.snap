// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MixerDetails Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Loading"
    data-testid="loading-indicator"
  />
</DocumentFragment>
`;
